### ISSUE TITLE
TypeError: Cannot read property 'linkReferences' of undefined

### ISSUE THREAD

**Issue by Bobface** [ethereum-optimism/plugins#13]
### Bug in the compiler plugin
When compiling an `abstract` contract which does not generate bytecode when compiled on its own results in an exception from the compiler plugin. It seems like the plugin is not handling that situation correctly.

Replacing 
https://github.com/ethereum-optimism/plugins/blob/f2cacfd4a41cc925fdb14f2141d59088b289ce67/src/hardhat/compiler/index.ts#L154

with
```
const linkRefs = contractOutput.evm.bytecode ? contractOutput.evm.bytecode.linkReferences : {};
```

makes the compilation work again.

### Example

Error:
```
PS C:\Users\user\Desktop\ideamarket-contracts> npx hardhat compile
Compiling 73 files with 0.6.12
Compiling 68 files with OVM compiler 0.6.12
An unexpected error occurred:

TypeError: Cannot read property 'linkReferences' of undefined
    at OverriddenTaskDefinition._action (C:\Users\user\Desktop\ideamarket-contracts\node_modules\@eth-optimism\plugins\src\hardhat\compiler\index.ts:160:56)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at Environment._runTaskDefinition (C:\Users\user\Desktop\ideamarket-contracts\node_modules\hardhat\src\internal\core\runtime-environment.ts:217:14)
    at Environment.run (C:\Users\user\Desktop\ideamarket-contracts\node_modules\hardhat\src\internal\core\runtime-environment.ts:130:14)
    at SimpleTaskDefinition.action (C:\Users\user\Desktop\ideamarket-contracts\node_modules\hardhat\src\builtin-tasks\compile.ts:661:18)
    at Environment._runTaskDefinition (C:\Users\user\Desktop\ideamarket-contracts\node_modules\hardhat\src\internal\core\runtime-environment.ts:217:14)
    at Environment.run (C:\Users\user\Desktop\ideamarket-contracts\node_modules\hardhat\src\internal\core\runtime-environment.ts:130:14)
    at Environment._runTaskDefinition (C:\Users\user\Desktop\ideamarket-contracts\node_modules\hardhat\src\internal\core\runtime-environment.ts:217:14)
    at Environment.run (C:\Users\user\Desktop\ideamarket-contracts\node_modules\hardhat\src\internal\core\runtime-environment.ts:130:14)
    at SimpleTaskDefinition.action (C:\Users\user\Desktop\ideamarket-contracts\node_modules\hardhat\src\builtin-tasks\compile.ts:945:37)
    at Environment._runTaskDefinition (C:\Users\user\Desktop\ideamarket-contracts\node_modules\hardhat\src\internal\core\runtime-environment.ts:217:14)
    at Environment.run (C:\Users\user\Desktop\ideamarket-contracts\node_modules\hardhat\src\internal\core\runtime-environment.ts:130:14)
    at SimpleTaskDefinition.action (C:\Users\user\Desktop\ideamarket-contracts\node_modules\hardhat\src\builtin-tasks\compile.ts:384:45)
    at Environment._runTaskDefinition (C:\Users\user\Desktop\ideamarket-contracts\node_modules\hardhat\src\internal\core\runtime-environment.ts:217:14)
    at Environment.run (C:\Users\user\Desktop\ideamarket-contracts\node_modules\hardhat\src\internal\core\runtime-environment.ts:130:14)
    at SimpleTaskDefinition.action (C:\Users\user\Desktop\ideamarket-contracts\node_modules\hardhat\src\builtin-tasks\compile.ts:1172:63)
```

Package versions:
``` 
> npm list --depth=0

+-- @eth-optimism/dev@1.1.1
+-- @eth-optimism/plugins@0.0.17
+-- @nomiclabs/hardhat-ethers@2.0.1
+-- @nomiclabs/hardhat-waffle@2.0.1
+-- @nomiclabs/hardhat-web3@2.0.0
+-- @openzeppelin/cli@2.8.2
+-- @openzeppelin/contracts@3.3.0
+-- @openzeppelin/test-helpers@0.5.9
+-- bignumber.js@9.0.1
+-- chai@4.3.3
+-- dotenv@8.2.0
+-- ethereum-waffle@3.3.0
+-- ethers@5.0.31
+-- hardhat@2.1.0
+-- moment@2.29.1
+-- prettier@2.1.2
+-- prompt@1.0.0
+-- solidity-coverage@0.7.13
`-- web3@1.3.4
```

Happens with both `@eth-optimism/plugins@0.0.16` and `@eth-optimism/plugins@0.0.17`. 

Adding some `console.log`s to `plugins/src/hardhat/compiler/index.js` shows the following values for the involved variables:

```
Compiling 75 files with 0.6.12

contractName: Context
fileName: @openzeppelin/contracts/GSN/Context.sol
contractOutput: { abi: [], evm: { methodIdentifiers: {} } }
An unexpected error occurred:

TypeError: Cannot read property 'linkReferences' of undefined
....
```

Here this bugs seems to happen with OZ's `Context` contract , which is `abstract`:
```
// SPDX-License-Identifier: MIT

pragma solidity >=0.6.0 <0.8.0;

/*
 * @dev Provides information about the current execution context, including the
 * sender of the transaction and its data. While these are generally available
 * via msg.sender and msg.data, they should not be accessed in such a direct
 * manner, since when dealing with GSN meta-transactions the account sending and
 * paying for execution may not be the actual sender (as far as an application
 * is concerned).
 *
 * This contract is only required for intermediate, library-like contracts.
 */
abstract contract Context {
    function _msgSender() internal view virtual returns (address payable) {
        return msg.sender;
    }

    function _msgData() internal view virtual returns (bytes memory) {
        this; // silence state mutability warning without generating bytecode - see https://github.com/ethereum/solidity/issues/2691
        return msg.data;
    }
}
```

### Reproduce

```
git clone https://github.com/ideamarket/ideamarket-contracts.git
cd ideamarket-contracts
git checkout dd1e90f
npm i
npx hardhat compile
```

**Comment by Bobface** [comment]
More info:
Adding some `console.log`s to `plugins/src/hardhat/compiler/index.js` shows the following values for the involved variables:

```
Compiling 75 files with 0.6.12

contractName: Context
fileName: @openzeppelin/contracts/GSN/Context.sol
contractOutput: { abi: [], evm: { methodIdentifiers: {} } }
An unexpected error occurred:

TypeError: Cannot read property 'linkReferences' of undefined
....
```

This bugs seems to happen with OZ's `Context` contract :
```
// SPDX-License-Identifier: MIT

pragma solidity >=0.6.0 <0.8.0;

/*
 * @dev Provides information about the current execution context, including the
 * sender of the transaction and its data. While these are generally available
 * via msg.sender and msg.data, they should not be accessed in such a direct
 * manner, since when dealing with GSN meta-transactions the account sending and
 * paying for execution may not be the actual sender (as far as an application
 * is concerned).
 *
 * This contract is only required for intermediate, library-like contracts.
 */
abstract contract Context {
    function _msgSender() internal view virtual returns (address payable) {
        return msg.sender;
    }

    function _msgData() internal view virtual returns (bytes memory) {
        this; // silence state mutability warning without generating bytecode - see https://github.com/ethereum/solidity/issues/2691
        return msg.data;
    }
}
```

**Comment by kfichter** [comment]
Just want to say that this is a fantastic bug report! I've gone ahead and made a PR in the plugins repo that should fix this issue (https://github.com/ethereum-optimism/plugins/pull/14). Once that gets merged I'll make a PR to release a new version.

**Comment by kfichter** [comment]
0.0.18 should be released soon

**Comment by Bobface** [comment]
Awesome, thank you!

**Comment by kfichter** [comment]
CI just published 0.0.18. I tested against your repo again and it should be working now. Please let me know if you have any more issues!

**Comment by kfichter** [comment]
Closing this, can reopen if you're still having the same problem.
